swarm/swap_test.go

```diff
@@ -0,0 +1,423 @@
+		if *printStats {
+			for k, v := range balancesMap {
+				fmt.Println(fmt.Sprintf("node %s balances:", k.TerminalString()))
+				for kk, vv := range v {
+					fmt.Println(fmt.Sprintf(".........with node %s: balance %d", kk.TerminalString(), vv))
+				}
+			}
+		}
+
+		//now iterate the whole map
+		//and check that every node k has the same
+		//balance with a peer as that peer with the node,
+		//but in inverted signs
+
+		//iterate the map
+		success := true
+		for k, mapForK := range balancesMap {
+			//iterate the submap
+			for n, balanceKwithN := range mapForK {
+				//iterate the main map again
+				for subK, mapForSubK := range balancesMap {
+					//if the node and the peer are the same...
+					if n == subK {
+						log.Trace(fmt.Sprintf("balance of %s with %s: %d", k.TerminalString(), n.TerminalString(), balanceKwithN))
+						log.Trace(fmt.Sprintf("balance of %s with %s: %d", n.TerminalString(), k.TerminalString(), mapForSubK[k]))
+						//...check that they have the same balance in Abs terms and that it is not 0
+						if balanceKwithN+mapForSubK[k] != 0 && balanceKwithN != 0 {
+							log.Error(fmt.Sprintf("Expected balances to be a+b = 0 AND balance(a) != 0, but they are not, balance k with n:  %d, balance n with k: %d", balanceKwithN, mapForSubK[k]))
+							success = false
+						}
+					}
+				}
+			}
+		}
+		if success {
+			return nil
+		}
+		return errors.New("some conditions could not be met")
+	})
+
+	if result.Error != nil {
+		t.Fatal(result.Error)
+	}
+	log.Debug("test terminated")
+}
+
+//TestSwapNetworkAsymmetricFileUpload is a swap test too,
+//but this time the number and size of files are random
+func TestSwapNetworkAsymmetricFileUpload(t *testing.T) {
+	nodeCount := 16
+
+	sim := simulation.New(map[string]simulation.ServiceFunc{
+		"swarm": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			config := api.NewConfig()
+			config.Port = strconv.Itoa(8500 + rand.Intn(9999))
+
+			dir, err := ioutil.TempDir("", "swap-network-test-node"+config.Port)
+			if err != nil {
+				return nil, nil, err
+			}
+			cleanup = func() {
+				err := os.RemoveAll(dir)
+				if err != nil {
+					log.Error("cleaning up swarm temp dir", "err", err)
+				}
+			}
+
+			config.Path = dir
+
+			privkey, err := crypto.GenerateKey()
+			if err != nil {
+				return nil, cleanup, err
+			}
+
+			config.Init(privkey)
+			//enable swap
+			config.SwapEnabled = true
+
+			swarm, err := NewSwarm(config, nil)
+			if err != nil {
+				return nil, cleanup, err
+			}
+			bucket.Store(bucketKeySwarm, swarm)
+			log.Info("new swarm", "bzzKey", config.BzzKey, "baseAddr", fmt.Sprintf("%x", swarm.bzz.BaseAddr()))
+			return swarm, cleanup, nil
+		},
+	})
+	defer sim.Close()
+
+	ctx := context.Background()
+	files := make([]file, 0)
+
+	var checkStatusM sync.Map
+	var nodeStatusM sync.Map
+	var totalFoundCount uint64
+
+	_, err := sim.AddNodesAndConnectChain(nodeCount)
```

Comment:
why in a chain and not full or use snapshot?

Reply:
Most of our tests connect in chain so I naively assumed this to be kind of default and sufficient.

Is it not? The tests pass though

Reply:
if you want a realistic end to end test, use a snapshot